### ISSUE TITLE
Terraform init is extremely slow

### ISSUE THREAD

**Issue by AlmogCohen** [hashicorp/terraform#27379]
Terraform init takes forever and it seems like it is due to provider download. This extremely slow behaviour is consistent for the last 2 days. Working like that is hellish as I'm using terragrunt which runs the `terraform init` command often in my case.

Drilling down on the trace logs, I've seen that the download of https://releases.hashicorp.com/terraform-provider-aws/3.22.0/terraform-provider-aws_3.22.0_darwin_amd64.zip takes usually 1-5 minutes minimum. When I simply download this url in my browser, it never takes more than 15 seconds....usually less than 10. When I run the `init` command with trace logs, it is almost *always* takes many many minutes to finish.

I have no internet issues at ALL <img width="645" alt="Screen Shot 2020-12-29 at 2 55 49 PM" src="https://user-images.githubusercontent.com/3888998/103285144-f3b50f80-49e5-11eb-8aa2-87e77c589fe5.png">

### Terraform Version
```
Terraform v0.13.5
+ provider registry.terraform.io/hashicorp/aws v3.22.0
```

### Terraform Configuration Files
I have created only one file in my example project, `backend.tf` and this is the its content
```# Generated by Terragrunt. Sig: nIlQXj57tbuaRZEa
terraform {
  backend "s3" {
    encrypt        = true
    key            = "staging/elastic_beanstalk/terraform.tfstate"
    profile        = "some-profile"
    region         = "us-east-1"
    bucket         = "company-terraform-state"
    dynamodb_table = "terraform-lock-table"
  }
}
```

### Debug Output
https://gist.github.com/AlmogCohen/b8af6e33c170c393a63e126226ad7108

### Crash Output
Not relevant

### Expected Behavior
Terraform provider download should take as fast as it takes in the browser, and terraform init should not take 5 minutes for a very very simple terraform proejct 
### Actual Behavior
Terraform provider/plugin downloads should be as fast as they execute in browser
### Steps to Reproduce
1. `terraform init`

### Additional Context
I'm using terragrunt, so the `init` command runs often on one of my dependency projects. It doesn't really matter for this bug report because I've run the `terraform` command directly in a folder which has only `tf` files in.

### References
Not relevant

**Comment by AlmogCohen** [comment]
Here is another trace log example from a project that is includes a little more pieces https://gist.github.com/AlmogCohen/19495e839bc4894d8084e9ab191e23f6

A billion of things are being downloaded, but mostly when it is requested from the hashicorp releases host, the download takes forever.

The download of `https://releases.hashicorp.com/terraform-provider-aws/2.70.0/terraform-provider-aws_2.70.0_darwin_amd64.zip` took almost a minute. Download of `https://releases.hashicorp.com/terraform-provider-template/2.2.0/terraform-provider-template_2.2.0_darwin_amd64.zip` took almost 40 seconds via the `terraform init` command, while via browser it took literally 2 seconds.

**Comment by danieldreier** [comment]
Thanks for reporting this! You're right that it shouldn't be so slow. I'm running on a mac and it's not that slow for me. Can you try and do an install of terraform via homebrew? Homebrew's Terraform has cgo enabled, and I'm wondering if you're stalling on DNS resolution.

**Comment by AlmogCohen** [comment]
Thanks! I appreciate your quick (and affirming) response!
I'm pretty sure I'm using the homebrew version as I've upgraded from
Terraform v0.12 a few days ago...the last homebrew command I have in
history is `brew reinstall terraform@0.13`

If I had DNS issues...won't it be affecting the browser download speed as
well? For all modules I tried, the browser speed was like 20 times faster
than the `terraform init` for that specific module.

**Comment by zzn01** [comment]
same issue here,  terrafrom version is 0.13.5, installed via homebrew 
hang on aws provider for very long time(10 minutes+)
```
- Installing -/aws v3.22.0...
```

**Comment by danieldreier** [comment]
@AlmogCohen the reason I wanted to look into DNS issues is that golang uses its own DNS resolver rather than the native one, and this has caused issues in Terraform (and many other projects) that can currently only be worked around with a cgo-enabled build. However, if you're not seeing it on a homebrew build then that's probably not the issue.

Most of the Terraform Core engineering team is out until next week, so I'm just doing basic triage of urgent issues. As other folks run into this - please upvote with a 👍 rather than a comment, to help us prioritize this. I don't think this will get looked at until the team is back but if it turns out there's a more widespread issue I can call folks in.

**Comment by danieldreier** [comment]
I went and tried to reproduce this again on my mac with the following trivial config:

```terraform
terraform {
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "~> 3.0"
    }
  }
}

# Configure the AWS Provider
provider "aws" {
  region = "us-east-1"
}
```

My init took about 7 seconds:
```
time terraform init

Initializing the backend...

Initializing provider plugins...
- Finding hashicorp/aws versions matching "~> 3.0"...
- Installing hashicorp/aws v3.22.0...
- Installed hashicorp/aws v3.22.0 (signed by HashiCorp)

Terraform has created a lock file .terraform.lock.hcl to record the provider
selections it made above. Include this file in your version control repository
so that Terraform can guarantee to make the same selections by default when
you run "terraform init" in the future.

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
terraform init  1.77s user 0.56s system 33% cpu 6.883 total
```

I do believe you that you're running into this, just wanted to clarify that I'm not able to reproduce a more widespread problem. 

Can you share a traceroute to releases.hashicorp.com? Okay to obscure the first few hops for privacy, I'm just trying to get clues on what's different between our environments. Also, can you share what OS version you're running, if it's X86 vs ARM, anything interesting about your network environment like an HTTP proxy, etc?

**Comment by AlmogCohen** [comment]
I copied your trivial config into `main.tf` and run `time terraform init` and this is the result:
```
Initializing the backend...

Initializing provider plugins...
- Finding hashicorp/aws versions matching "~> 3.0"...
- Installing hashicorp/aws v3.22.0...
- Installed hashicorp/aws v3.22.0 (signed by HashiCorp)

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
terraform init  2.73s user 1.57s system 7% cpu 1:00.33 total
```

Just to prove the point again regarding no internet issues, I've run `time wget https://releases.hashicorp.com/terraform-provider-aws/3.22.0/terraform-provider-aws_3.22.0_darwin_amd64.zip` and got this:
```
--2020-12-31 08:19:33--  https://releases.hashicorp.com/terraform-provider-aws/3.22.0/terraform-provider-aws_3.22.0_darwin_amd64.zip
Resolving releases.hashicorp.com (releases.hashicorp.com)... 151.101.1.183, 151.101.129.183, 151.101.65.183, ...
Connecting to releases.hashicorp.com (releases.hashicorp.com)|151.101.1.183|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 42005468 (40M) [application/zip]
Saving to: ‘terraform-provider-aws_3.22.0_darwin_amd64.zip’

terraform-provider-aws_3.22.0_darwin_amd64.zip     100%[================================================================================================================>]  40.06M  10.1MB/s    in 4.7s

2020-12-31 08:19:38 (8.61 MB/s) - ‘terraform-provider-aws_3.22.0_darwin_amd64.zip’ saved [42005468/42005468]

wget   0.10s user 0.25s system 6% cpu 5.137 total
```

Mac specifications: 
<img width="252" alt="Screen Shot 2020-12-31 at 8 25 07 AM" src="https://user-images.githubusercontent.com/3888998/103397592-b1a0e080-4b41-11eb-96ab-c230b29733a4.png">

For the `traceroute`, I think some Firewall drops it midway because nothing can be seen after the first few hops:
```
traceroute -q 1 -w 1 -m 100 releases.hashicorp.com
traceroute: Warning: releases.hashicorp.com has multiple addresses; using 151.101.1.183
traceroute to dualstack.s.shared.global.fastly.net (151.101.1.183), 100 hops max, 52 byte packets
 1   #### Removed #####
 2   #### Removed #####
 3  *
 4   #### Removed #####
 5  #### Removed #####
 6  *
 7  bb302-pt-e49-100g-hot.israelinternet.co.il (185.149.252.109)  11.057 ms
 8  10.141.226.60 (10.141.226.60)  12.209 ms
 9  10.185.180.51 (10.185.180.51)  15.525 ms
10  10.185.180.11 (10.185.180.11)  12.616 ms
11  be6037.203.rcr22.fra06.atlas.cogentco.com (149.14.210.161)  69.101 ms
12  *
13 *
.......(nothing answered here....)
97  *
98  *
99  *
100  *
```

**Comment by pselle** [comment]
@AlmogCohen Thanks for running the timings and providing additional data.`terraform init` will take longer than the GET request alone.

One suggestion I have if you're running init often (not a workflow that is currently optimized), do make sure that if you can you're using [provider cache settings](https://www.terraform.io/docs/commands/cli-config.html#provider-plugin-cache). This should help with regard to your workflow.

On the final comments with timing, I see @danieldreier's report with ~7 seconds, and yours with ~5 seconds ... neither of these are close to the original 1-5 minutes in the original report. @danieldreier was also correct to point to CGO builds, as there are known DNS issues/traps that occur with other versions.

Would enabling the cache help your workflow?

**Comment by AlmogCohen** [comment]
@pselle Thanks for your comment. I'll look into enabling the cache later this week and respond here.

Regardless, it is important to note that the original problem still stands in place. The timing was *1 minute* (and not 5 seconds like you said). If you'll take a closer look, you can see that the 5s timing is of *wget*. Terraform init is still 12 times slower, and from previous logs you could see that 95% of the time spent on terraform init is due to the resource download of the aws provider (which was timed as 5s via wget).

**Comment by pselle** [comment]
@AlmogCohen Thanks, you're correct, I was looking at the wrong timing! It was the 1 minute.

**Comment by future-media** [comment]
I have the same problem on Windows. Checking with Wireshark, I can see that the releases.hashicorp.com server starts off by accepting a TLSv1.2 connection but then gets stuck returning messages saying "ignored unknown record". It took so long that I killed it.

I was just doing the learn-terraform-aws-instance tutorial. I was able to create infrastructure once, but never again after that.

**Comment by future-media** [comment]
Although  learn-terraform-aws-instance tutorial fails completely on my office connection as described above, it works just fine when executed from Windows Server hosted by a different provider.

**Comment by lmserrano** [comment]
Something that also aggravates this issue is the fact that apparently, if you run multiple consecutive `terraform init commands without any changes in the used providers nor their versions, Terraform will still proceed to install them each time:

```bash
$ terraform init
Initializing modules...

Initializing the backend...

Initializing provider plugins...
- Reusing previous version of hashicorp/random from the dependency lock file
- Reusing previous version of hashicorp/local from the dependency lock file
- Reusing previous version of hashicorp/null from the dependency lock file
- Reusing previous version of hashicorp/template from the dependency lock file
...
- Installing hashicorp/template v2.2.0...
- Installed hashicorp/template v2.2.0 (signed by HashiCorp)
- Installing hashicorp/random v3.0.0...
- Installed hashicorp/random v3.0.0 (signed by HashiCorp)
- Installing hashicorp/local v2.0.0...
- Installed hashicorp/local v2.0.0 (signed by HashiCorp)
- Installing hashicorp/null v3.0.0...
- Installed hashicorp/null v3.0.0 (signed by HashiCorp)
...

Terraform has been successfully initialized!
```

Enabling cache may improve the speed of this operation, but it will still take a while because in Terraform `0.14.4` it is doing all this each time, while in previous versions `< 0.13.x (if I'm not mistaken)` it would only download what was new or that it wouldn't have, or providers which had new minor versions in case their version numbers were referenced with `~>`.

**Comment by debu99** [comment]
this is a really bad experience when creating modules

**Comment by HsnVahedi** [comment]
I'm using terraform 0.14.5 on ubuntu. I'm experiencing the same thing. `terraform init` takes a lot of time. here is my` main.tf`:

```
provider "kubernetes" {
}

resource "kubernetes_deployment" "backend" {
  metadata {
    name = "backend"
    labels = {
      app = "backend"
    }
  }

  spec {
    replicas = 3

    selector {
      match_labels = {
        app = "backend"
      }
    }

    template {
      metadata {
        labels = {
          app = "backend"
        }
      }

      spec {
        container {
          image = "IMAGE"
          name  = "backend"
          port {
            container_port = 8000
          }
        }
      }
    }
  }
}
```

**Comment by gevial** [comment]
Init takes 5 minutes in our setup, that's quite frustrating :(
Terraform 0.13.3

**Comment by lmserrano** [comment]
@gevial , if you can, try updating to 0.14.6. It has a nice set of improvements related with init, especially for what concerns default caching config.

**Comment by kalyanar4** [comment]
for me the terraform Init takes forever, do we have a fix for this?

**Comment by future-media** [comment]
I never got it to go, so I log into an instance of Windows Server on a different network and Terraform from there.

**Comment by martavoi** [comment]
It has just stuck on my mac:
```
Initializing the backend...

Initializing provider plugins...
- Finding hashicorp/aws versions matching "~> 3.27"...
- Installing hashicorp/aws v3.38.0...
```

```
terraform {
  backend "s3" {
    bucket = "boss-mr-terraform-state-dev"
    key    = "obs"
    dynamodb_table = "terraform-lock"
    region = "us-east-1"
  }
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "~> 3.27"
    }
  }

  required_version = ">= 0.14.9"
}

provider "aws" {
  region  = "us-east-1"
}
```

**Comment by lsascha** [comment]
for me as well most of the time.
could be that the server where the providers are hosted is slow or some other networking issue.

I already enabled the plugin cache, but unfortunately i think gitlab-runner running in kubernetes do not have a shared cache.

**Comment by scastria** [comment]
Yes, it is now extremely slow for me now too.  It never used to be.  I am also on a mac.

**Comment by lkishalmi** [comment]
Could it be an IPv6 issue?
I'm experiencing this over T-Mobile 5g as my primary ISP is down.

**Comment by future-media** [comment]
Yes, possibly it is related to IPv6. My office connection where it fails is IPv6 and my server where it works is IPv4.

**Comment by scastria** [comment]
I found the issue for me.  My corporate IT installed a network security product a few weeks ago called Zscaler. It had not caused any trouble before so I didn’t think about it. However, I also noticed than when I would watch streaming video lately that I could never get an HD stream, only low quality streams. I disabled Zscaler and it fixed both issues, video streaming and TF init plug-in downloads.

**Comment by marceloboeira** [comment]
Any updates here? having the same issues... downloading from providers is extremely slow while everything else network-wise seems fine...

**Comment by lyc0221** [comment]
Any updates here? having the same issues...

**Comment by moarychan** [comment]
If there's a mirror in China for developers will be very nice, turn on the VPN, but the bellow error message is often encountered.
```text
terraform init

Initializing the backend...

Initializing provider plugins...
- Finding hashicorp/azurerm versions matching ">= 2.75.0"...
- Finding aztfmod/azurecaf versions matching "1.2.10"...
- Installing hashicorp/azurerm v2.93.1...
╷
│ Error: Failed to install provider
│
│ Error while installing hashicorp/azurerm v2.93.1: Get
│ "https://releases.hashicorp.com/terraform-provider-azurerm/2.93.1/terraform-provider-azurerm_2.93.1_linux_amd64.zip": dial tcp 23.235.32.32:443: i/o
│ timeout
```

**Comment by CrazyOptimist** [comment]
Same issue here, it sometimes takes more than 5 minutes on my laptop, which has 100MB/s download speed.

**Comment by lots0logs** [comment]
A year later and this issue still exists in latest version of terraform. Can someone provide an update?? 

![bump](https://user-images.githubusercontent.com/4675662/153973452-4e0e17a4-6ecb-4283-957b-b4ece1daa1b4.gif)

**Comment by Ripeey** [comment]
Same, it was all working fine until today all of sudden it take me over 5 minutes to "init".

**Comment by rdonkin-attest** [comment]
Getting the same issue from UK with Terraform 1.1.7:

- client is macOS 11.6, home firewall + FTTP router connection - 100 Mbps

This took 8 minutes to upgrade providers for an existing module:

```
$ terraform init -upgrade
Upgrading modules...
- xxxx in modules/vpn

Initializing the backend...

Initializing provider plugins...
- Finding hashicorp/aws versions matching "~> 3.74"...
- Finding hashicorp/kubernetes versions matching "1.13.3"...
- Finding latest version of hashicorp/template...
- Installing hashicorp/template v2.2.0...
- Installed hashicorp/template v2.2.0 (signed by HashiCorp)
- Installing hashicorp/aws v3.75.1...
- Installed hashicorp/aws v3.75.1 (signed by HashiCorp)
- Installing hashicorp/kubernetes v1.13.3...
- Installed hashicorp/kubernetes v1.13.3 (signed by HashiCorp)

Terraform has made some changes to the provider dependency selections recorded
in the .terraform.lock.hcl file. Review those changes and commit them to your
version control system if they represent changes you intended to make.

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
```

I turned off IPv6 and tested on a single ethernet interface only with WiFi disabled.   Traceroute was very slow (15 minutes to complete 64 hops, and all hops after 8 were `* * *`):

```
$ traceroute -n releases.hashicorp.com
traceroute to dualstack.s.shared.global.fastly.net (199.232.25.183), 64 hops max, 52 byte packets
...
 4  185.66.207.2  9.105 ms  9.250 ms  9.255 ms
 5  * 216.66.86.149  10.579 ms *
 6  184.104.192.133  9.850 ms * *
 7  72.52.92.198  20.624 ms  20.538 ms
    72.52.92.202  15.547 ms
 8  * * *
 9  * * *
10  * * *
11  * * *
12  * * *
```

However, this seems to be an artefact of traceroute - the similar third party `mtr` (in Homebrew) runs a number of traceroutes and shows much faster results.  However it may be using a different technique:

```
$ sudo mtr -c 10 --report releases.hashicorp.com
Start: 2022-04-12T14:51:46+0100
HOST: xxxxxxxxxxx                  Loss%   Snt   Last   Avg  Best  Wrst StDev
  1.|-- fw                         0.0%    10    0.5   0.7   0.5   0.9   0.1
  2.|-- 192.168.1.1                0.0%    10    1.0   1.1   0.9   1.7   0.2
  3.|-- 100.64.88.1                0.0%    10    9.2   9.5   9.2   9.9   0.2
  4.|-- 185.66.207.2               0.0%    10    9.2   9.5   9.2  10.0   0.3
  5.|-- e0-17.core3.lon1.he.net   50.0%    10   10.4  10.6  10.0  11.7   0.7
  6.|-- 100ge0-33.core2.lon3.he.n 70.0%    10   10.3  10.7  10.3  11.0   0.4
  7.|-- ip81-59.fastly-gw1.lonap.  0.0%    10    9.4   9.6   9.3  10.0   0.2
  8.|-- 151.101.1.183              0.0%    10    9.5   9.6   9.0  10.2   0.3
```

Not sure what's going on here, and previous terraform inits have been fine (with same networking setup since late Dec 2021).

I hope someone from HashiCorp can help debug this -  I don't use any other apps/tools with this level of delay, and the `mtr` results suggest it's not a layer 3 network issue.

**Comment by madsonic** [comment]
Same issue from Thailand. on 0.15.x

**Comment by CCCCCaO** [comment]
same issue from China on 1.0.11 with windows10

**Comment by everdark** [comment]
same issue from Singapore on Apple M1 (macOS Monterey 12.2)

The provider is as simple as:

```tf
terraform {
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "4.0.0"
    }
  }
}

provider "aws" {
  region = "ap-southeast-1"
}
```

and hang at

```
Initializing provider plugins...
- Finding hashicorp/aws versions matching "4.0.0"...
- Installing hashicorp/aws v4.0.0...
```

**Comment by kentwongg** [comment]
So is HashiCorp going to comment?

**Comment by libracoder** [comment]
Having same issue

**Comment by m18unet** [comment]
I'm having the same problem. Any progress?

**Comment by simbu21** [comment]
I m having the same issue. Any update. Thank you.

**Comment by pintoflager** [comment]
Terraform v1.2.2
on linux_amd64

...and soooo sloooooow

**Comment by jaliagag** [comment]
Same issue - trying [caching](https://www.terraform.io/cli/config/config-file#provider-plugin-cache), as [recommended above](https://github.com/hashicorp/terraform/issues/27379#issuecomment-754100075) 
We'll see how it works

**Comment by binhoul** [comment]
Same issue...

**Comment by mrnonz** [comment]
Same issue with Terraform v1.2.5 

Interim solution is manual remove `.terraform` folder and run `terraform init` again.

**Comment by crw** [comment]
@jaliagag Did caching help?

@mrnonz if you remove the `.terraform` folder, it completes successfully and within a reasonable amount of time? What happens if you run it again afterwards, with that folder?

**Comment by jeffery** [comment]
Had the same issue on my macOS Monterey (12.5). Disabled ipv6 to `Link-local only` and restarted my mac which resolved the issue for me.

**Comment by lsascha** [comment]
only that it is not always possible to change network settings, like in a CI-CD Environment.

**Comment by Kaydub00** [comment]
I'm also experiencing this, @jeffery fix seemed to make it faster, but it's still pretty slow to download providers

**Comment by ChristophShyper** [comment]
I run Terraform via Terragrunt inside a Docker container. Worked perfectly for years, until M2 Macs come around.
On Intel processor typical run is 2 minutes. On M2 it's 5 minutes. Those 3 minutes are spent waiting on Terraform initialization.
Tested on MacBook Pro M2 and MacBook Air M2. Disabling IPv6 does nothing.

**Comment by apparentlymart** [comment]
Hi all,

It isn't really clear how we can make progress on this issue, because there is no clear signal about what is causing these problems. We've seen several different theories discussed above, including:
* Not resolving DNS records correctly on macOS, leading to problems for folks using VPNs or other kinds of split-horizon DNS. 
 
    That was true of older versions of Terraform, but the `darwin_amd64` builds have been using correct resolver for a long time now and we just released v1.3.1 which includes the same fix for `darwin_arm64` (Apple Silicon) builds. Therefore I think we can safely exclude this as a possible cause for modern Terraform.
* Internet service providers which have a small MTU for packets, therefore requiring the remote server to throttle back its packet sizes in response to ICMP notifications.

    I only see one mention of this and I don't have such a system to test with, so this one is still on the table as a possible cause but I'm not sure how to decide whether it's an _actual_ cause. This is a tricky one because it's a hard constraint of the internet connection and so not something that can be temporarily disabled to see if it changes the outcome.
* Some of you are using networks with filewalls or other interfering middleboxes, and have found that Terraform works better with those disabled.

    if we can narrow down a specific thing that Terraform is doing that irritates these intermediaries which isn't true of other network software then we could potentially try to address it, but these systems are typically pretty opaque about how they do their work and each works differently. We don't have direct access to any and so we'd need to rely on those who use these products to let us know exactly what Terraform is doing that violates the assumptions of these filters.
* IPV6 problems various: it sounds like some of you have IPv6 connectivity that is measurably slower than your IPv4 connectivity, and perhaps Terraform was choosing to use IPv6 and therefore ending up with a slower transfer speed than other software.

    Some time ago we changed Terraform to use [the Fast Fallback algorithm](https://en.wikipedia.org/wiki/Happy_Eyeballs), which aims to avoid problems like this by trying both IPv4 and IPv6 simultaneously and using IPv4 if the IPv6 connection doesn't work. However, I believe that only helps if the IPv6 connection is totally unusable, and not if IPv6 is working but slower than IPv4.

As before, we believe you all that Terraform is working more slowly than some other network software when run on your computer but I'm at a loss as to what to do about it unless someone can either describe exactly why Terraform is slow for their particular situation or can describe a situation we can reproduce on our own systems to debug it ourselves. Our team is distributed across the US, Canada, and Europe with different ISPs each, but we've not been able to reproduce these problems on our systems or on the cloud platforms we have access to.

If you _are_ able to download provider packages more effectively using software other than Terraform itself, you may benefit from setting up a local mirror of the providers you use regularly on your computer, and then setting up [a custom provider installation strategy](https://www.terraform.io/cli/config/config-file#provider-installation). That means you can in principle download the provider `.zip` packages from `releases.hashicorp.com` or from GitHub releases (depending on the provider) using your browser or other software and arrange them into one of the local filesystem mirror directory structures, and then Terraform will just load the provider plugins from your local disk rather than trying to install them over the network. This if course only a workaround and not a solution, but in the absence of concrete information about what causes the slow transfer rates I don't really have anything else to suggest.

If anyone does have any evidence of specific network misbehavior in Terraform, please let us know!

**Comment by AlmogCohen** [comment]
@apparentlymart your thorough response is very much appreciated! 

For the protocol, as the original issue reporter....years have passed and I've since upgraded my MacOS, changed my address, my ISP and the terraform versions I use (now always >1.0.0 for all projects). I haven't seen the issue I reported  for years now...

The fact that "it works on my machine" does not invalidate the many other reports that came in and seem to be related to the very same "stage" of the terraform init. On the other hand, I do understand how challenging it may be to take any action here to resolve the (many?) issues.

**Comment by PSanetra** [comment]
I try to install terraform providers locally on my machine (Linux, Manjaro). Installing them is currently really slow (I usually do not use or at least initialize terraform locally, but run it from a CI, where some providers also might be cached).

I notice that slow performance only with terraform. E.g. pulling Docker images from Docker Hub is fast as usual.

I grabbed a sample provider plugin from the terraform logs and tried to download it via chrome. Here are the screenshots:
```
[...]
- Installing hashicorp/azurerm v2.70.0...
2023-02-08T14:34:27.971+0100 [TRACE] providercache.Dir.InstallPackage: installing registry.terraform.io/hashicorp/azurerm v2.70.0 from https://releases.hashicorp.com/terraform-provider-azurerm/2.70.0/terraform-provider-azurerm_2.70.0_linux_amd64.zip
2023-02-08T14:34:27.971+0100 [TRACE] HTTP client GET request to https://releases.hashicorp.com/terraform-provider-azurerm/2.70.0/terraform-provider-azurerm_2.70.0_linux_amd64.zip
```
![slow_terraform_provider_download](https://user-images.githubusercontent.com/7327745/217546952-c6acc2b6-724a-4e23-95ce-19ad410c6956.png)

If my location is relevant: I am located in Germany.

**Comment by nibblesnbits** [comment]
Same issue. I guess #32863 is a duplicate of this.

**Comment by queglay** [comment]
I think terraform init is slow even when its working!  
It adds 10 seconds every time its executed in a minimal AWS infra test example of mine and it needs to be executed every time in CI.  It would be good if it had some serialisation of versions to determine if it needs to do heavy lifting or just use the existing cache and be snappier.

**Comment by jsyrjala** [comment]
Below command took over 46 minutes. The slow part was `Fetching hashicorp/aws 5.26.0`.

Using MacOS. The only network related thing that is slow on this machine is Terraform.

```
time terraform providers lock -platform=darwin_amd64 -platform=darwin_arm64 -platform=linux_amd64 -platform=windows_amd64

- Fetching hashicorp/aws 5.26.0 for darwin_amd64...
- Retrieved hashicorp/aws 5.26.0 for darwin_amd64 (signed by HashiCorp)
- Fetching hashicorp/random 3.5.1 for darwin_amd64...
- Retrieved hashicorp/random 3.5.1 for darwin_amd64 (signed by HashiCorp)
- Fetching hashicorp/random 3.5.1 for darwin_arm64...
- Retrieved hashicorp/random 3.5.1 for darwin_arm64 (signed by HashiCorp)
- Fetching hashicorp/aws 5.26.0 for darwin_arm64...
- Retrieved hashicorp/aws 5.26.0 for darwin_arm64 (signed by HashiCorp)
- Fetching hashicorp/aws 5.26.0 for linux_amd64...
- Retrieved hashicorp/aws 5.26.0 for linux_amd64 (signed by HashiCorp)
- Fetching hashicorp/random 3.5.1 for linux_amd64...
- Retrieved hashicorp/random 3.5.1 for linux_amd64 (signed by HashiCorp)
- Fetching hashicorp/aws 5.26.0 for windows_amd64...
- Retrieved hashicorp/aws 5.26.0 for windows_amd64 (signed by HashiCorp)
- Fetching hashicorp/random 3.5.1 for windows_amd64...
- Retrieved hashicorp/random 3.5.1 for windows_amd64 (signed by HashiCorp)
- Obtained hashicorp/aws checksums for darwin_amd64; All checksums for this platform were already tracked in the lock file
- Obtained hashicorp/aws checksums for darwin_arm64; Additional checksums for this platform are now tracked in the lock file
- Obtained hashicorp/aws checksums for linux_amd64; Additional checksums for this platform are now tracked in the lock file
- Obtained hashicorp/aws checksums for windows_amd64; Additional checksums for this platform are now tracked in the lock file
- Obtained hashicorp/random checksums for darwin_amd64; All checksums for this platform were already tracked in the lock file
- Obtained hashicorp/random checksums for darwin_arm64; All checksums for this platform were already tracked in the lock file
- Obtained hashicorp/random checksums for linux_amd64; Additional checksums for this platform are now tracked in the lock file
- Obtained hashicorp/random checksums for windows_amd64; Additional checksums for this platform are now tracked in the lock file

Success! Terraform has updated the lock file.

Review the changes in .terraform.lock.hcl and then commit to your
version control system to retain the new checksums.

terraform providers lock -platform=darwin_amd64 -platform=darwin_arm64    22.81s user 8.99s system 1% cpu 46:02.31 total
```

**Comment by jsyrjala** [comment]
Seems that https://releases.hashicorp.com/ is some times extremely slow for downloading.

```
➜  tmp wget https://releases.hashicorp.com/terraform-provider-template/2.2.0/terraform-provider-template_2.2.0_darwin_amd64.zip
--2023-11-20 14:12:36--  https://releases.hashicorp.com/terraform-provider-template/2.2.0/terraform-provider-template_2.2.0_darwin_amd64.zip
Resolving releases.hashicorp.com (releases.hashicorp.com)... 18.165.140.5, 18.165.140.94, 18.165.140.83, ...
Connecting to releases.hashicorp.com (releases.hashicorp.com)|18.165.140.5|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 7438707 (7,1M) [application/zip]
Saving to: ‘terraform-provider-template_2.2.0_darwin_amd64.zip’

terraform-provider-template_2.2.0_darwin_amd64.zip         100%[=====================================================================================================================================>]   7,09M  13,7MB/s    in 0,5s    

2023-11-20 14:12:37 (13,7 MB/s) - ‘terraform-provider-template_2.2.0_darwin_amd64.zip’ saved [7438707/7438707]

➜  tmp wget https://releases.hashicorp.com/terraform-provider-aws/5.26.0/terraform-provider-template_5.26.0_darwin_amd64.zip
--2023-11-20 14:13:03--  https://releases.hashicorp.com/terraform-provider-aws/5.26.0/terraform-provider-template_5.26.0_darwin_amd64.zip
Resolving releases.hashicorp.com (releases.hashicorp.com)... 18.165.140.5, 18.165.140.94, 18.165.140.83, ...
Connecting to releases.hashicorp.com (releases.hashicorp.com)|18.165.140.5|:443... connected.
HTTP request sent, awaiting response... 404 Not Found
2023-11-20 14:13:04 ERROR 404: Not Found.

➜  tmp wget https://releases.hashicorp.com/terraform-provider-aws/5.26.0/terraform-provider-aws_5.26.0_darwin_arm64.zip     
--2023-11-20 14:13:31--  https://releases.hashicorp.com/terraform-provider-aws/5.26.0/terraform-provider-aws_5.26.0_darwin_arm64.zip
Resolving releases.hashicorp.com (releases.hashicorp.com)... 18.165.140.5, 18.165.140.94, 18.165.140.83, ...
Connecting to releases.hashicorp.com (releases.hashicorp.com)|18.165.140.5|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 85950151 (82M) [application/zip]
Saving to: ‘terraform-provider-aws_5.26.0_darwin_arm64.zip’

terraform-provider-aws_5.26.0_darwin_arm64.zip             100%[=====================================================================================================================================>]  81,97M  18,5MB/s    in 4,6s    

2023-11-20 14:13:35 (17,9 MB/s) - ‘terraform-provider-aws_5.26.0_darwin_arm64.zip’ saved [85950151/85950151]

➜  tmp wget https://releases.hashicorp.com/terraform-provider-aws/5.26.0/terraform-provider-aws_5.26.0_darwin_arm64.zip
--2023-11-20 14:13:49--  https://releases.hashicorp.com/terraform-provider-aws/5.26.0/terraform-provider-aws_5.26.0_darwin_arm64.zip
Resolving releases.hashicorp.com (releases.hashicorp.com)... 18.165.140.5, 18.165.140.94, 18.165.140.83, ...
Connecting to releases.hashicorp.com (releases.hashicorp.com)|18.165.140.5|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 85950151 (82M) [application/zip]
Saving to: ‘terraform-provider-aws_5.26.0_darwin_arm64.zip.1’

terraform-provider-aws_5.26.0_darwin_arm64.zip.1             5%[======>                                                                                                                               ]   4,32M  34,4KB/s    eta 11m 51s^C
➜  tmp wget https://releases.hashicorp.com/terraform-provider-aws/5.26.0/terraform-provider-aws_5.26.0_darwin_arm64.zip
--2023-11-20 14:14:32--  https://releases.hashicorp.com/terraform-provider-aws/5.26.0/terraform-provider-aws_5.26.0_darwin_arm64.zip
Resolving releases.hashicorp.com (releases.hashicorp.com)... 18.165.140.94, 18.165.140.83, 18.165.140.50, ...
Connecting to releases.hashicorp.com (releases.hashicorp.com)|18.165.140.94|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 85950151 (82M) [application/zip]
Saving to: ‘terraform-provider-aws_5.26.0_darwin_arm64.zip.2’

terraform-provider-aws_5.26.0_darwin_arm64.zip.2             7%[=========>                                                                                                                            ]   6,25M  29,0KB/s    eta 12m 41s
```

**Comment by zhenrong-wang** [comment]
> I think terraform init is slow even when its working!  
> It adds 10 seconds every time its executed in a minimal AWS infra test example of mine and it needs to be executed every time in CI.  It would be good if it had some serialisation of versions to determine if it needs to do heavy lifting or just use the existing cache and be snappier.

From my experience, only the first time of init takes time to download (cache) the provider binary. As long as the binary got downloaded and cached to local environment, init will take effect instantly. 

One solution might be:

Download  and unzip all the provider binary(ies) to your local terraform provider registry. 

Reference URL:  https://releases.hashicorp.com

Most dev providers can be downloaded from their github repositories.

**Comment by queglay** [comment]
Perhaps theres something I'm not getting. run locally it is fast, and reuses the lock file I have committed.  
```
user@MacBook-Pro-2 tf-aws-lambda-nebula % terraform init

Initializing Terraform Cloud...
Initializing modules...

Initializing provider plugins...
- Reusing previous version of hashicorp/archive from the dependency lock file
- Reusing previous version of hashicorp/aws from the dependency lock file
- Using previously-installed hashicorp/archive v2.4.0
- Using previously-installed hashicorp/aws v4.4.0

Terraform Cloud has been successfully initialized!
```


but my circle ci runner + terraform cloud runs always have to do more:
```
Initializing Terraform Cloud...
Initializing modules...
- bucket_policy_attachments in modules/s3-shared-bucket-policy
- certificates_bucket in modules/s3-nebula-bucket
- scripts_bucket in modules/s3-nebula-bucket

Initializing provider plugins...
- Reusing previous version of hashicorp/aws from the dependency lock file
- Reusing previous version of hashicorp/archive from the dependency lock file
- Installing hashicorp/aws v4.4.0...
- Installed hashicorp/aws v4.4.0 (signed by HashiCorp)
- Installing hashicorp/archive v2.4.0...
- Installed hashicorp/archive v2.4.0 (signed by HashiCorp)

Terraform has made some changes to the provider dependency selections recorded
in the .terraform.lock.hcl file. Review those changes and commit them to your
version control system if they represent changes you intended to make.

Terraform Cloud has been successfully initialized!
```

**Comment by queglay** [comment]
Ok I took a look at fixing this - it was due to my caching not setup properly.  but still takes about 6 seconds here when on my mac (speed test is about 108MBit/s from Melbourne Australia):
Initializing Terraform Cloud...

init in cicrcle ci is now 1-2 secs though with caching fixed so that is nice to know.

**Comment by swiknaba** [comment]
For me, `terraform init` downloads files at a whopping speed of 150 kb/s. When I download anything else via the terminal (e.g. `brew install...` or `docker pull`) I get around 50 MBit.

Terraform is consistently slow for me, it's a nightmare.

**Comment by wes-nimblerx** [comment]
I think it's something to do with https://registry.terraform.io
My system hangs on a variety of URLs, mostly during afternoon time PST

Maybe the below will help
```
Downloading registry.terraform.io/cloudposse/label/null 0.25.0 for alb.default_load_balancer_label...
2025-01-22T13:38:03.807-0800 [DEBUG] looking up module location from "https://registry.terraform.io/v1/modules/cloudposse/label/null/0.25.0/download"
2025-01-22T13:38:03.807-0800 [DEBUG] GET https://registry.terraform.io/v1/modules/cloudposse/label/null/0.25.0/download
2025-01-22T13:38:58.525-0800 [DEBUG] Module installer: alb.default_load_balancer_label installed at .terraform/modules/alb.default_load_balancer_label
- alb.default_load_balancer_label in .terraform/modules/alb.default_load_balancer_label
2025-01-22T13:38:58.525-0800 [DEBUG] Module installer: begin alb.default_target_group_label
Downloading registry.terraform.io/cloudposse/label/null 0.25.0 for alb.default_target_group_label...
2025-01-22T13:38:58.532-0800 [DEBUG] Module installer: alb.default_target_group_label installed at .terraform/modules/alb.default_target_group_label
- alb.default_target_group_label in .terraform/modules/alb.default_target_group_label
2025-01-22T13:38:58.532-0800 [DEBUG] Module installer: begin alb.this
Downloading registry.terraform.io/cloudposse/label/null 0.25.0 for alb.this...
2025-01-22T13:38:58.540-0800 [DEBUG] Module installer: alb.this installed at .terraform/modules/alb.this
- alb.this in .terraform/modules/alb.this
╷
│ Error: Failed to download module
│ 
│   on .terraform/modules/alb/main.tf line 53:
│   53: module "access_logs" {
│ 
│ Could not download module "access_logs" (.terraform/modules/alb/main.tf:53) source code from
│ "git::https://github.com/cloudposse/terraform-aws-lb-s3-bucket?ref=502f2f75fb91730cbf986c55e9560ad4163e5c12": error downloading
│ 'https://github.com/cloudposse/terraform-aws-lb-s3-bucket?ref=502f2f75fb91730cbf986c55e9560ad4163e5c12': /opt/homebrew/bin/git exited with 128: Cloning into
│ '.terraform/modules/alb.access_logs'...
│ error: RPC failed; curl 92 HTTP/2 stream 5 was not closed cleanly: CANCEL (err 8)
│ error: 14937 bytes of body are still expected
│ fetch-pack: unexpected disconnect while reading sideband packet
│ fatal: early EOF
│ fatal: fetch-pack: invalid index-pack output
│ .
```

The slowness happens on a variety of URLs.  When it hangs in another terminal it downloads fast
```
$ time curl https://registry.terraform.io/v1/modules/cloudposse/alb/aws/2.2.1/download

real	0m0.175s
$ time curl  https://registry.terraform.io/v1/modules/cloudposse/lb-s3-bucket/aws/0.19.0/download

real	0m0.121s
$ time curl https://registry.terraform.io/v1/modules/cloudposse/s3-bucket/aws/3.1.2/download

real	0m0.215s
$ time curl https://registry.terraform.io/v1/modules/cloudposse/iam-system-user/aws/1.0.0/download

real	0m0.225s
```
Here is a snippet of the hanging times it took to download from 30 to 150 seconds stalls
```
$ grep -A1 "GET https://registry.terraform.io" /tmp/t
2025-01-22T13:42:46.741-0800 [DEBUG] GET https://registry.terraform.io/v1/modules/cloudposse/alb/aws/versions
2025-01-22T13:42:46.761-0800 [DEBUG] found available version "0.1.0" for cloudposse/alb/aws
--
2025-01-22T13:42:46.762-0800 [DEBUG] GET https://registry.terraform.io/v1/modules/cloudposse/alb/aws/2.2.1/download
2025-01-22T13:43:29.485-0800 [DEBUG] Module installer: alb installed at .terraform/modules/alb
--
2025-01-22T13:43:29.485-0800 [DEBUG] GET https://registry.terraform.io/v1/modules/cloudposse/lb-s3-bucket/aws/versions
2025-01-22T13:43:29.510-0800 [DEBUG] found available version "0.1.0" for cloudposse/lb-s3-bucket/aws
--
2025-01-22T13:43:29.510-0800 [DEBUG] GET https://registry.terraform.io/v1/modules/cloudposse/lb-s3-bucket/aws/0.19.0/download
2025-01-22T13:43:31.194-0800 [DEBUG] Module installer: alb.access_logs installed at .terraform/modules/alb.access_logs
--
2025-01-22T13:43:31.194-0800 [DEBUG] GET https://registry.terraform.io/v1/modules/cloudposse/s3-log-storage/aws/versions
2025-01-22T13:43:31.212-0800 [DEBUG] found available version "0.1.0" for cloudposse/s3-log-storage/aws
--
2025-01-22T13:43:31.212-0800 [DEBUG] GET https://registry.terraform.io/v1/modules/cloudposse/s3-log-storage/aws/1.4.2/download
2025-01-22T13:43:32.181-0800 [DEBUG] Module installer: alb.access_logs.s3_bucket installed at .terraform/modules/alb.access_logs.s3_bucket
--
2025-01-22T13:43:32.181-0800 [DEBUG] GET https://registry.terraform.io/v1/modules/cloudposse/s3-bucket/aws/versions
2025-01-22T13:43:32.200-0800 [DEBUG] found available version "0.1.0" for cloudposse/s3-bucket/aws
--
2025-01-22T13:43:32.200-0800 [DEBUG] GET https://registry.terraform.io/v1/modules/cloudposse/s3-bucket/aws/3.1.2/download
2025-01-22T13:46:03.263-0800 [DEBUG] Module installer: alb.access_logs.s3_bucket.aws_s3_bucket installed at .terraform/modules/alb.access_logs.s3_bucket.aws_s3_bucket
--
2025-01-22T13:46:03.264-0800 [DEBUG] GET https://registry.terraform.io/v1/modules/cloudposse/iam-s3-user/aws/versions
2025-01-22T13:46:03.360-0800 [DEBUG] found available version "0.1.0" for cloudposse/iam-s3-user/aws
--
2025-01-22T13:46:03.360-0800 [DEBUG] GET https://registry.terraform.io/v1/modules/cloudposse/iam-s3-user/aws/1.1.0/download
2025-01-22T13:46:04.690-0800 [DEBUG] Module installer: alb.access_logs.s3_bucket.aws_s3_bucket.s3_user installed at .terraform/modules/alb.access_logs.s3_bucket.aws_s3_bucket.s3_user
--
2025-01-22T13:46:04.690-0800 [DEBUG] GET https://registry.terraform.io/v1/modules/cloudposse/iam-system-user/aws/versions
2025-01-22T13:46:04.708-0800 [DEBUG] found available version "0.1.0" for cloudposse/iam-system-user/aws
--
2025-01-22T13:46:04.709-0800 [DEBUG] GET https://registry.terraform.io/v1/modules/cloudposse/iam-system-user/aws/1.0.0/download
2025-01-22T13:46:44.487-0800 [DEBUG] Module installer: alb.access_logs.s3_bucket.aws_s3_bucket.s3_user.s3_user installed at .terraform/modules/alb.access_logs.s3_bucket.aws_s3_bucket.s3_user.s3_user
--
2025-01-22T13:46:44.487-0800 [DEBUG] GET https://registry.terraform.io/v1/modules/cloudposse/ssm-parameter-store/aws/versions
2025-01-22T13:46:44.504-0800 [DEBUG] found available version "0.1.0" for cloudposse/ssm-parameter-store/aws
--
2025-01-22T13:46:44.505-0800 [DEBUG] GET https://registry.terraform.io/v1/modules/cloudposse/ssm-parameter-store/aws/0.10.0/download
2025-01-22T13:47:35.332-0800 [DEBUG] Module installer: alb.access_logs.s3_bucket.aws_s3_bucket.s3_user.s3_user.store_write installed at .terraform/modules/alb.access_logs.s3_bucket.aws_s3_bucket.s3_user.s3_user.store_write
--
2025-01-22T13:47:35.332-0800 [DEBUG] GET https://registry.terraform.io/v1/modules/cloudposse/label/null/versions
2025-01-22T13:47:35.351-0800 [DEBUG] found available version "0.1.0" for cloudposse/label/null
--
2025-01-22T13:47:35.351-0800 [DEBUG] GET https://registry.terraform.io/v1/modules/cloudposse/label/null/0.25.0/download
2025-01-22T13:49:57.155-0800 [DEBUG] Module installer: alb.access_logs.s3_bucket.aws_s3_bucket.s3_user.s3_user.store_write.this installed at .terraform/modules/alb.access_logs.s3_bucket.aws_s3_bucket.s3_user.s3_user.store_write.this
--
```

**Comment by wes-nimblerx** [comment]
Unfortunately the plugin-cache does not cache the cloudposse modules.  This does help, but does not solve the issue.
This is what gets downloaded
```
Downloading registry.terraform.io/cloudposse/ecs-container-definition/aws 0.61.1 for container_definition...
Downloading git::https://github.com/nimblerx/terraform-aws-ecs-web-app.git?ref=main for ecs_web...
Downloading registry.terraform.io/cloudposse/alb-ingress/aws 0.28.0 for ecs_web.alb_ingress...
Downloading registry.terraform.io/cloudposse/label/null 0.25.0 for ecs_web.alb_ingress.target_group...
Downloading registry.terraform.io/cloudposse/label/null 0.25.0 for ecs_web.alb_ingress.this...
Downloading registry.terraform.io/cloudposse/alb-target-group-cloudwatch-sns-alarms/aws 0.17.0 for ecs_web.alb_target_group_cloudwatch_sns_alarms...
Downloading registry.terraform.io/cloudposse/label/null 0.25.0 for ecs_web.alb_target_group_cloudwatch_sns_alarms.httpcode_elb_5xx_alarm_label...
Downloading registry.terraform.io/cloudposse/label/null 0.25.0 for ecs_web.alb_target_group_cloudwatch_sns_alarms.httpcode_target_3xx_alarm_label...
Downloading registry.terraform.io/cloudposse/label/null 0.25.0 for ecs_web.alb_target_group_cloudwatch_sns_alarms.httpcode_target_4xx_alarm_label...
Downloading registry.terraform.io/cloudposse/label/null 0.25.0 for ecs_web.alb_target_group_cloudwatch_sns_alarms.httpcode_target_5xx_alarm_label...
Downloading registry.terraform.io/cloudposse/label/null 0.25.0 for ecs_web.alb_target_group_cloudwatch_sns_alarms.target_response_time_alarm_label...
Downloading registry.terraform.io/cloudposse/label/null 0.25.0 for ecs_web.alb_target_group_cloudwatch_sns_alarms.this...
Downloading registry.terraform.io/cloudposse/ecs-container-definition/aws 0.58.1 for ecs_web.container_definition...
Downloading registry.terraform.io/cloudposse/ecr/aws 0.41.0 for ecs_web.ecr...
Downloading registry.terraform.io/cloudposse/label/null 0.25.0 for ecs_web.ecr.this...
Downloading git::https://github.com/nimblerx/terraform-aws-ecs-alb-service-task.git?ref=main for ecs_web.ecs_alb_service_task...
Downloading registry.terraform.io/cloudposse/label/null 0.25.0 for ecs_web.ecs_alb_service_task.exec_label...
Downloading registry.terraform.io/cloudposse/label/null 0.25.0 for ecs_web.ecs_alb_service_task.service_connect_label...
Downloading registry.terraform.io/cloudposse/label/null 0.25.0 for ecs_web.ecs_alb_service_task.service_label...
Downloading registry.terraform.io/cloudposse/label/null 0.25.0 for ecs_web.ecs_alb_service_task.task_label...
Downloading registry.terraform.io/cloudposse/label/null 0.25.0 for ecs_web.ecs_alb_service_task.this...
Downloading registry.terraform.io/cloudposse/ecs-cloudwatch-autoscaling/aws 0.7.5 for ecs_web.ecs_cloudwatch_autoscaling...
Downloading registry.terraform.io/cloudposse/label/null 0.25.0 for ecs_web.ecs_cloudwatch_autoscaling.scale_down_label...
Downloading registry.terraform.io/cloudposse/label/null 0.25.0 for ecs_web.ecs_cloudwatch_autoscaling.scale_up_label...
Downloading registry.terraform.io/cloudposse/label/null 0.25.0 for ecs_web.ecs_cloudwatch_autoscaling.this...
Downloading registry.terraform.io/cloudposse/ecs-cloudwatch-sns-alarms/aws 0.12.2 for ecs_web.ecs_cloudwatch_sns_alarms...
Downloading registry.terraform.io/cloudposse/label/null 0.25.0 for ecs_web.ecs_cloudwatch_sns_alarms.cpu_utilization_high_alarm_label...
Downloading registry.terraform.io/cloudposse/label/null 0.25.0 for ecs_web.ecs_cloudwatch_sns_alarms.cpu_utilization_low_alarm_label...
Downloading registry.terraform.io/cloudposse/label/null 0.25.0 for ecs_web.ecs_cloudwatch_sns_alarms.memory_utilization_high_alarm_label...
Downloading registry.terraform.io/cloudposse/label/null 0.25.0 for ecs_web.ecs_cloudwatch_sns_alarms.memory_utilization_low_alarm_label...
Downloading registry.terraform.io/cloudposse/label/null 0.24.1 for ecs_web.ecs_cloudwatch_sns_alarms.this...
Downloading registry.terraform.io/cloudposse/ecs-codepipeline/aws 0.34.1 for ecs_web.ecs_codepipeline...
Downloading registry.terraform.io/cloudposse/codebuild/aws 2.0.2 for ecs_web.ecs_codepipeline.codebuild...
Downloading registry.terraform.io/cloudposse/label/null 0.25.0 for ecs_web.ecs_codepipeline.codebuild.this...
Downloading registry.terraform.io/cloudposse/label/null 0.25.0 for ecs_web.ecs_codepipeline.codebuild_label...
Downloading registry.terraform.io/cloudposse/label/null 0.25.0 for ecs_web.ecs_codepipeline.codepipeline_assume_role_label...
Downloading registry.terraform.io/cloudposse/label/null 0.25.0 for ecs_web.ecs_codepipeline.codepipeline_label...
Downloading registry.terraform.io/cloudposse/label/null 0.25.0 for ecs_web.ecs_codepipeline.codepipeline_s3_policy_label...
Downloading registry.terraform.io/cloudposse/label/null 0.25.0 for ecs_web.ecs_codepipeline.codestar_label...
Downloading registry.terraform.io/cloudposse/repository-webhooks/github 0.13.0 for ecs_web.ecs_codepipeline.github_webhooks...
Downloading registry.terraform.io/cloudposse/label/null 0.25.0 for ecs_web.ecs_codepipeline.github_webhooks.this...
Downloading registry.terraform.io/cloudposse/label/null 0.25.0 for ecs_web.ecs_codepipeline.this...
Downloading registry.terraform.io/cloudposse/label/null 0.25.0 for ecs_web.this...
Downloading registry.terraform.io/cloudposse/iam-role/aws 0.20.0 for role...
Downloading registry.terraform.io/cloudposse/label/null 0.25.0 for role.role_name...
Downloading registry.terraform.io/cloudposse/label/null 0.25.0 for role.this...
Downloading registry.terraform.io/cloudposse/security-group/aws 2.2.0 for sg...
Downloading registry.terraform.io/cloudposse/label/null 0.25.0 for sg.this...
```

This is what got cached
```
$ find .terraform.d -type d
.terraform.d
.terraform.d/plugin-cache
.terraform.d/plugin-cache/registry.terraform.io
.terraform.d/plugin-cache/registry.terraform.io/hashicorp
.terraform.d/plugin-cache/registry.terraform.io/hashicorp/local
.terraform.d/plugin-cache/registry.terraform.io/hashicorp/local/2.5.2
.terraform.d/plugin-cache/registry.terraform.io/hashicorp/local/2.5.2/darwin_arm64
.terraform.d/plugin-cache/registry.terraform.io/hashicorp/null
.terraform.d/plugin-cache/registry.terraform.io/hashicorp/null/3.2.3
.terraform.d/plugin-cache/registry.terraform.io/hashicorp/null/3.2.3/darwin_arm64
.terraform.d/plugin-cache/registry.terraform.io/hashicorp/aws
.terraform.d/plugin-cache/registry.terraform.io/hashicorp/aws/5.80.0
.terraform.d/plugin-cache/registry.terraform.io/hashicorp/aws/5.80.0/darwin_arm64
.terraform.d/plugin-cache/registry.terraform.io/hashicorp/random
.terraform.d/plugin-cache/registry.terraform.io/hashicorp/random/3.6.3
.terraform.d/plugin-cache/registry.terraform.io/hashicorp/random/3.6.3/darwin_arm64
.terraform.d/plugin-cache/registry.terraform.io/integrations
.terraform.d/plugin-cache/registry.terraform.io/integrations/github
.terraform.d/plugin-cache/registry.terraform.io/integrations/github/6.4.0
.terraform.d/plugin-cache/registry.terraform.io/integrations/github/6.4.0/darwin_arm64
```

**Comment by apparentlymart** [comment]
Hi all,

I don't work on Terraform anymore but since I previously commented here I keep getting notified about the new comments 😀 and noticed that the discussion has slightly shifted since my previous comment made while I was still a Terraform team member. None of the following should be understood as an official statement since I'm now just some random guy leaving comments on GitHub!

At the point I left my earlier comment we'd been talking about situations where Terraform seemed to be making its own network requests in some different way to other software on the same computer, which therefore appeared to cause performance problems due to unique characteristics of the particular computer or network where Terraform was running. I'd said that we'd not been able to identify any specific cause and therefore it wasn't clear how to proceed.

However, the recent comments seem to be making the more general observation that it's slow to install lots of remote modules from the Terraform registry. That seems like quite a different situation: in that case Terraform is literally just running `git clone` on a bunch of repositories in a loop, and so the slowness there seems caused not by unique network conditions but just by the fact that cloning a git repository over the network just _takes time_, and the more modules you add the more time it takes.

That _particular_ observation seems already captured better by https://github.com/hashicorp/terraform/issues/28299, and has a proposed solution in https://github.com/hashicorp/terraform/issues/34634. Although neither of those issues seem to address it directly, another relevant observation is that Terraform could potentially try to install many providers and modules concurrently rather than installing one at a time sequentially, which would potentially speed things up even without adding a "module cache".

I don't know what the Terraform team's intentions are with _this_ issue, but I can say that my intention at the time of my earlier comment was to put this one on pause until we could actually find some specific information on what Terraform was doing differently to other network software that was causing the _raw data transfer_ to be slower than it ought to be, and assuming that remains true (which it might not) it might be better to pursue ideas for having `terraform init` just _go faster in general_ (regardless of specific quirks on certain computers and OSes) in those other issues I linked.

**Comment by jbardin** [comment]
As stated in the previous comment, there don't seem to be any more reports resembling the original issue here. Since we have other issues tracking possible optimizations for various init subsystems, I'm going to close this particular issue out.
